packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -1466,8 +1466,10 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
       total: body.hits.total,
       saved_objects: body.hits.hits.map(
         (hit: estypes.SearchHit<SavedObjectsRawDocSource>): SavedObjectsFindResult => ({
-          // @ts-expect-error @elastic/elasticsearch _source is optional
-          ...this._rawToSavedObject(hit, { migrationVersionCompatibility }),
+          ...(this._migrator.migrateDocument(
+            // @ts-expect-error @elastic/elasticsearch _source is optional
+            this._rawToSavedObject(hit, { migrationVersionCompatibility })
+          ) as SavedObject),
```

Comment:
yeah when we planned this work I missed the fact that an update to a type with deferred migrations could cause data loss. Since update doesn't set the version, it could also cause us to call the migration twice.

```
// Document
{
  panelsJSON: "...v1PanelsJson..."
  typeMigrationVersion: 1
}

// On-read we apply deferred migration
{
  panelsJSON: "...v2PanelsJson..."
  typeMigrationVersion: 2
}

// If we then update just the panelsJSON attribute we have a document like the following in the index
{
  panelsJSON: "...v2PanelsJson..."
  typeMigrationVersion: 1 // <-- wrong version
}

// If we read the updated document we end up applying the up transform on `v2PanelsJson` as if it was actually still a v1PanelsJson
```

I don't expect it would be obvious to other teams. So I wonder if we should merge this but not expose it just yet (hide the types). We're anyway planning on solving the update problem for ZDT so once that's in place we can enable this.

Reply:
Yeah, we will likely need some or all of the mechanism for zdt later on. I would be fine merging without enablement I think